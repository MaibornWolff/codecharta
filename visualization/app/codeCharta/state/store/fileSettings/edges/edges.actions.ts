import { CCAction, Edge } from "../../../../codeCharta.model"

export enum EdgesActions {
	SET_EDGES = "SET_EDGES",
	ADD_EDGE = "ADD_EDGE",
	REMOVE_EDGE = "REMOVE_EDGE"
}

export interface SetEdgesAction extends CCAction {
	type: EdgesActions.SET_EDGES
	payload: Edge[]
}

export interface AddEdgeAction extends CCAction {
	type: EdgesActions.ADD_EDGE
	payload: Edge
}

export interface RemoveEdgeAction extends CCAction {
	type: EdgesActions.REMOVE_EDGE
	payload: Edge
}

export type EdgesAction = SetEdgesAction | AddEdgeAction | RemoveEdgeAction

export function setEdges(edges: Edge[] = defaultEdges): SetEdgesAction {
	return {
		type: EdgesActions.SET_EDGES,
		payload: edges
	}
}

export function addEdge(edge: Edge): AddEdgeAction {
	return {
		type: EdgesActions.ADD_EDGE,
		payload: edge
	}
}

export function removeEdge(edge: Edge): RemoveEdgeAction {
	return {
		type: EdgesActions.REMOVE_EDGE,
		payload: edge
	}
}

export const defaultEdges: Edge[] = []
