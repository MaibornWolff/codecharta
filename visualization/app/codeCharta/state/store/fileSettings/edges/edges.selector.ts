import { createSelector } from "@ngrx/store"
import { fileSettingsSelector } from "../fileSettings.selector"

export const edgesSelector = createSelector(fileSettingsSelector, fileSettings => fileSettings.edges)
