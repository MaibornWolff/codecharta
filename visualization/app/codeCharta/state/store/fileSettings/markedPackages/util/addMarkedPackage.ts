import { MarkedPackage } from "../../../../../codeCharta.model"
import { getParent } from "../../../../../util/nodePathHelper"

export const addMarkedPackage = (markedPackagesMap: Map<string, MarkedPackage>, { path, color }: MarkedPackage) => {
	const directMarkedParentPackage = getParent(markedPackagesMap, path)

	if (!directMarkedParentPackage || directMarkedParentPackage.color !== color) {
		markedPackagesMap.set(path, {
			path,
			color
		})
	}

	for (const [key, markedPackage] of markedPackagesMap) {
		if (markedPackage.path === path) {
			if (markedPackage.color !== color) {
				markedPackagesMap.delete(key)
			}
		} else if (markedPackage.path.startsWith(path)) {
			// Remove marked packages with color identical to their parent marked package
			const markedPackageParent = getParent(markedPackagesMap, markedPackage.path)
			if (markedPackageParent && markedPackageParent.color === markedPackage.color) {
				markedPackagesMap.delete(key)
			}
		}
	}
}
