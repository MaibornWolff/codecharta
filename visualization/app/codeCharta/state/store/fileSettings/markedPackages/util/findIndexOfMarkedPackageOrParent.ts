import { MarkedPackage } from "../../../../../codeCharta.model"

/** @return index of exact match if there is an exact match. Otherwise return index of parent or -1  */
export const findIndexOfMarkedPackageOrParent = (markedPackages: MarkedPackage[], nodePath: string) => {
	const indexOfNodePath = markedPackages.findIndex(mp => mp.path === nodePath)
	if (indexOfNodePath !== -1) return indexOfNodePath

	let indexOfNearestParent = -1
	for (let loopIndex = 0; loopIndex < markedPackages.length; loopIndex++) {
		const markedPackage = markedPackages[loopIndex]
		if (
			nodePath.startsWith(markedPackage.path) &&
			(indexOfNearestParent === -1 || markedPackages[indexOfNearestParent].path.length < markedPackage.path.length)
		)
			indexOfNearestParent = loopIndex
	}

	return indexOfNearestParent
}
