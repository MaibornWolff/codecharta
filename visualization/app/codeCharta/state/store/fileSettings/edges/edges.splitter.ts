import { setEdges } from "./edges.actions"
import { Edge } from "../../../../codeCharta.model"

export function splitEdgesAction(payload: Edge[]) {
	return setEdges(payload)
}
