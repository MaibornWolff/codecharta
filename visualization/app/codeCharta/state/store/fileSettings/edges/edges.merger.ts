import { CCFile, Edge } from "../../../../codeCharta.model"
import { getUpdatedPath } from "../../../../util/nodePathHelper"

export function getMergedEdges(inputFiles: CCFile[], withUpdatedPath: boolean): Edge[] {
	const edges: Edge[] = []

	if (inputFiles.length == 1) {
		return inputFiles[0].settings.fileSettings.edges
	}

	for (let inputFile of inputFiles) {
		if (inputFile.settings.fileSettings.edges) {
			for (let oldEdge of inputFile.settings.fileSettings.edges) {
				const edge: Edge = {
					fromNodeName: withUpdatedPath
						? getUpdatedPath(inputFile.fileMeta.fileName, oldEdge.fromNodeName)
						: oldEdge.fromNodeName,
					toNodeName: withUpdatedPath ? getUpdatedPath(inputFile.fileMeta.fileName, oldEdge.toNodeName) : oldEdge.toNodeName,
					attributes: oldEdge.attributes,
					visible: oldEdge.visible
				}
				const equalEdgeItem = edges.find(e => e.fromNodeName == edge.fromNodeName && e.toNodeName == edge.toNodeName)

				if (equalEdgeItem) {
					for (let key in edge.attributes) {
						equalEdgeItem.attributes[key] = edge.attributes[key]
					}
				} else {
					edges.push(edge)
				}
			}
		}
	}

	return edges
}
