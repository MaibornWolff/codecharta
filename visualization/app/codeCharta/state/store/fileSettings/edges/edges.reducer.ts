import { EdgesAction, EdgesActions, setEdges } from "./edges.actions"
import { Edge } from "../../../../codeCharta.model"
import { addItemToArray, removeItemFromArray } from "../../../../util/reduxHelper"
const clone = require("rfdc")()

export function edges(state: Edge[] = setEdges().payload, action: EdgesAction): Edge[] {
	switch (action.type) {
		case EdgesActions.SET_EDGES:
			return clone(action.payload)
		case EdgesActions.ADD_EDGE:
			return addItemToArray(state, action.payload)
		case EdgesActions.REMOVE_EDGE:
			return removeItemFromArray(state, action.payload)
		default:
			return state
	}
}
