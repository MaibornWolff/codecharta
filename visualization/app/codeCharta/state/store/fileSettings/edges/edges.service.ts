import { StoreService, StoreSubscriber } from "../../../store.service"
import { IRootScopeService } from "angular"
import { EdgesActions, setEdges } from "./edges.actions"
import _ from "lodash"
import { Edge, FileState } from "../../../../codeCharta.model"
import { FileStateService, FileStateSubscriber } from "../../../fileState.service"
import { getMergedEdges } from "./edges.merger"
import { FileStateHelper } from "../../../../util/fileStateHelper"

export interface EdgesSubscriber {
	onEdgesChanged(edges: Edge[])
}

export class EdgesService implements StoreSubscriber, FileStateSubscriber {
	private static EDGES_CHANGED_EVENT = "edges-changed"

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		StoreService.subscribe(this.$rootScope, this)
		FileStateService.subscribe(this.$rootScope, this)
	}

	public onStoreChanged(actionType: string) {
		if (_.values(EdgesActions).includes(actionType)) {
			this.notify(this.select())
		}
	}

	public onFileStatesChanged(fileStates: FileState[]) {
		this.merge(fileStates)
	}

	private merge(fileStates: FileState[]) {
		const visibleFiles = FileStateHelper.getVisibleFileStates(fileStates).map(x => x.file)
		const withUpdatedPath = FileStateHelper.isPartialState(fileStates)
		const newEdges = getMergedEdges(visibleFiles, withUpdatedPath)
		this.storeService.dispatch(setEdges(newEdges))
	}

	private select() {
		return this.storeService.getState().fileSettings.edges
	}

	private notify(newState: Edge[]) {
		this.$rootScope.$broadcast(EdgesService.EDGES_CHANGED_EVENT, { edges: newState })
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: EdgesSubscriber) {
		$rootScope.$on(EdgesService.EDGES_CHANGED_EVENT, (event, data) => {
			subscriber.onEdgesChanged(data.edges)
		})
	}
}
