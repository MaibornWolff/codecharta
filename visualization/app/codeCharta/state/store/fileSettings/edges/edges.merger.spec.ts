import { CCFile, Edge } from "../../../../codeCharta.model"
import { getMergedEdges } from "./edges.merger"
import { TEST_FILE_DATA } from "../../../../util/dataMocks"
import _ from "lodash"

describe("EdgesMerger", () => {
	describe("getMergedEdges", () => {
		let edge1: Edge
		let edge2: Edge
		let edge3: Edge
		let edge4: Edge

		let file1: CCFile
		let file2: CCFile

		beforeEach(() => {
			file1 = _.cloneDeep(TEST_FILE_DATA)
			file1.fileMeta.fileName = "file1"
			file1.settings.fileSettings.edges = []

			file2 = _.cloneDeep(TEST_FILE_DATA)
			file2.fileMeta.fileName = "file2"
			file2.settings.fileSettings.edges = []

			edge1 = {
				fromNodeName: "/root/nodeA",
				toNodeName: "/root/nodeB",
				attributes: {
					attribute1: 10,
					attribute2: 20
				}
			}

			edge2 = {
				fromNodeName: "/root/nodeA",
				toNodeName: "/root/nodeC",
				attributes: {
					attribute1: 10,
					attribute2: 20
				}
			}

			edge3 = {
				fromNodeName: "/root/nodeA",
				toNodeName: "/root/nodeB",
				attributes: {
					attribute3: 30,
					attribute4: 40
				}
			}

			edge4 = {
				fromNodeName: "/root/nodeA",
				toNodeName: "/root/nodeB",
				attributes: {
					attribute1: 70,
					attribute2: 80
				}
			}
		})

		it("should merge empty edges-arrays", () => {
			file1.settings.fileSettings.edges = []
			file2.settings.fileSettings.edges = []
			expect(getMergedEdges([file1, file2], false)).toEqual([])
		})

		it("should merge different edges", () => {
			file1.settings.fileSettings.edges = [edge1]
			file2.settings.fileSettings.edges = [edge2]
			expect(getMergedEdges([file1, file2], false)).toMatchSnapshot()
		})

		it("should merge all edges if one file does not contain edges", () => {
			file1.settings.fileSettings.edges = [edge1, edge2]
			file2.settings.fileSettings.edges = null
			expect(getMergedEdges([file1, file2], false)).toMatchSnapshot()
		})

		it("should merge edge-attributes for the same edge paths", () => {
			file1.settings.fileSettings.edges = [edge1]
			file2.settings.fileSettings.edges = [edge3]
			expect(getMergedEdges([file1, file2], false)).toMatchSnapshot()
		})

		it("should overwrite duplicated edge-attributes for the same edge", () => {
			file1.settings.fileSettings.edges = [edge1]
			file2.settings.fileSettings.edges = [edge4]
			expect(getMergedEdges([file1, file2], false)).toMatchSnapshot()
		})
	})
})
