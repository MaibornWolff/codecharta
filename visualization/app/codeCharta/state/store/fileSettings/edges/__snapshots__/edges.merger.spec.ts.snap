// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgesMerger getMergedEdges should merge all edges if one file does not contain edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeC",
    "visible": undefined,
  },
]
`;

exports[`EdgesMerger getMergedEdges should merge different edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeC",
    "visible": undefined,
  },
]
`;

exports[`EdgesMerger getMergedEdges should merge edge-attributes for the same edge paths 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 10,
      "attribute2": 20,
      "attribute3": 30,
      "attribute4": 40,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
]
`;

exports[`EdgesMerger getMergedEdges should overwrite duplicated edge-attributes for the same edge 1`] = `
Array [
  Object {
    "attributes": Object {
      "attribute1": 70,
      "attribute2": 80,
    },
    "fromNodeName": "/root/nodeA",
    "toNodeName": "/root/nodeB",
    "visible": undefined,
  },
]
`;